rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,708 @@
+
+If fields are missing from ScheduledCar, it is not known who is scheduled to operate that car. An empty object is valid if we know the car is scheduled to operate but don't know anything else about it.
+
+### Signature
+Represents the operator's signature in Glides confirming that they were fit for duty.
+
+Fields
+- `type` (string): How the the operator made their signature.
+
+Current values for `type` are
+- `"tap"` if they signed in by tapping their badge to an RFID scanner.
+- `"type"` if they signed in by typing their badge number into Glides as a signature.
+
+Represented as an object so we could include other types of signatures or other data about the signature in the future without a breaking change.
+
+The tapped badge's RFID serial number is not included in the stream for security reasons (it could be used to impersonate an employee).
+
+### Time
+Time in the `HH:MM:SS` format. The time is measured from "noon minus 12h" of the service day (effectively midnight except for days on which daylight savings time changes occur). For times occurring after midnight, enter the time as a value greater than 24:00:00 in `HH:MM:SS` local time for the day on which the trip schedule begins.  Effectively, a [GTFS Time](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#field-types) but without support for `H:MM:SS` times.
+> _Example: `04:30:00` for 4:30AM or `25:35:00` for 1:35AM on the next day._
+
+### TripAdded
+There is a new trip, that does not appear in the Glides' schedule data. "Added" is relative to the schedule that Glides uses, and an added trip may correspond to a trip that appears in a different schedule.
+
+Any added trips SHOULD appear exactly once a TripAdded object, and any future updates SHOULD appear as TripUpdated objects, (except in case of a duplicate event due to a network error, in which case the two events SHOULD have the same TripKey and data). Clients SHOULD ignore TripAdded events with a TripKey that has already been added.
+
+The event SHOULD only include values which were explicitly included by the author: inferred values SHOULD NOT be included.
+
+TripAdded objects are the same as TripUpdated, with the following changes, so that the new trip can be placed in the appropriate order within existing trips:
+
+New fields:
+- `nextTripKey` ([TripKey](#TripKey), optional): links the newly created trip to the trip after it. If the TripKey refers to an added trip, it may not have been seen in the event stream at the time of this event. Required if `startLocation` is specified and `startTime` is NOT specified.
+- `previousTripKey` ([TripKey](#TripKey), optional): links the newly created trip to the trip immediately before it. If the TripKey refers to an added trip, it may not have been seen in the event stream at the time of this event. Required if `endLocation` is specified and `endTime` is NOT specified.
```

Comment:
This isn't backwards. It's based on the way inspectors add round trips to Glides. We know the terminal that the trip starts and and the return trip ends at, and we know the start time of the first trip. But we have to infer the station at the other end, and the approximate times for first trip's endTime and the return trips' startTime and endTime.

In order to avoid added trips that are completely floating, unspecified in time, trips either need to have one of their times specified, or be before/after a trip that has its time specified.

So, when an inspector adds a round trip, and we send a TripAdded event for the return trip, we know the endStation, but don't know the endTime. We then require a `previousTripKey` to link it to the first trip of the round trip, which does have a startTime, so that consumers can guess approximately when the trip will happen.

---

But your comment did make me realize that the requirements here might not be right. Perhaps it'd be better as "either one of previousTripKey and nextTripKey are required if neither startTime nor endTime are set". As it is, the requirements prohibit a one-way added trip with a specified startTime, a missing endTime (to be inferred), and both locations set, which although it's not possible in Glides right now, is a totally reasonable trip to want to specify.

We should also consider requiring that the next/prev TripKey points to a trip with a set time, so that Glides can't send two trips without times pointing to each other, and having no way to infer the times. Though that might be hard to formally require, given that the other trip "may not have been seen in the event stream at the time of this event".

I also think that the requirements in the list below *are* backwards.

I guess this section still needs some work. Thanks for the comment.

Reply:
Okay, I simplified and fixed this section in https://github.com/mbta/technology-docs/pull/18/commits/daec911c2e5c7a68c9ac35411a864fe47938cd5c
- The requirement is now "at least one of `startTime`, `endTime`, and `previousTripKey` is required." This is all that's needed to be sure that the trip has a time, or can have a time inferred from the previous trip.
- Removed `nextTripKey`.
  - Justification: You shouldn't have to infer a trip's time based on a future trip, and having `previousTripKey` conditionally required and useful for times while `nextTripKey` isn't would have been confusing. Also, while it could be nice to have a bi-directional link between previous and next trips, we couldn't actually guarantee it, so don't want the field's presence to imply that a consumer could count on it.

Reply:
This seems to disallow a round-trip arriving at a given time?

Reply:
Yeah, it does disallow that, but that's intentional. Consumers shouldn't be expected to infer times about a past trip based on a future trip, inspectors wouldn't plan it in that way, and Glides doesn't have an interface for it.